File: android-lib/src/com/kinvey/android/AndroidCredentialStore.java
/*
 * Copyright (c) 2013 Kinvey Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package com.kinvey.android;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.AsyncTask;
import android.util.Log;
import com.google.common.base.Preconditions;

import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.util.HashMap;

import com.kinvey.java.auth.Credential;
import com.kinvey.java.auth.CredentialStore;

/**
 * @author mjsalinger
 * @since 2.0
 */
class AndroidCredentialStore implements CredentialStore {
    private static final String TAG = AndroidCredentialStore.class.getSimpleName();

    SharedPreferences preferences;
    HashMap<String, Credential> credentials;
    Context appContext;

    AndroidCredentialStore(Context context) throws IOException, AndroidCredentialStoreException {
        appContext = context.getApplicationContext();
        credentials = new HashMap<String, Credential>();
        try {
            retrieveCredentialStore();
        } catch (ClassNotFoundException ex) {
            credentials = new HashMap<String, Credential>();
            new PersistCredentialStore().execute();
            throw new AndroidCredentialStoreException("Credential store corrupted and was rebuilt");
        }
    }

    /** {@inheritDoc} */
    @Override
    public Credential load(String userId) throws IOException {
        return credentials.get(userId);
    }

    /** {@inheritDoc} */
    @Override
    public void store(String userId, Credential credential) throws IOException {
        Preconditions.checkNotNull(credential, "credential must not be null");
        Preconditions.checkNotNull(userId, "userId must not be null");

        credentials.put(userId, credential);
        new PersistCredentialStore().execute();
    }

    /** {@inheritDoc} */
    @Override
    public void delete(String userId) {
        credentials.remove(userId);
        new PersistCredentialStore().execute();
    }

    private void retrieveCredentialStore() throws ClassNotFoundException {
        FileInputStream fIn = null;
        ObjectInputStream in = null;

        if (appContext.getFileStreamPath("kinveyCredentials.bin").exists()) {
            try {
                fIn = appContext.openFileInput("kinveyCredentials.bin");
                in = new ObjectInputStream(fIn);
                credentials = (HashMap<String, Credential>) in.readObject();
            } catch (IOException ex) {
                Log.w(TAG, "Corrupt credential store detected", ex);
            } finally {
                try {
                    if (fIn != null) {
                        fIn.close();
                    }

                    if (in != null) {
                        in.close();
                    }

                } catch (IOException ioe) {
                    Log.w(TAG, "Could not clean up resources", ioe);
                }
            }
        } else {
            new PersistCredentialStore().execute();
        }
    }

    private class PersistCredentialStore extends AsyncTask<Void, Void, Void> {

        @Override
        protected Void doInBackground(Void... voids) {
            try {
                FileOutputStream fStream = appContext.openFileOutput("kinveyCredentials.bin", Context.MODE_PRIVATE);
                ObjectOutputStream oStream = new ObjectOutputStream(fStream);

                oStream.writeObject(credentials);
                oStream.flush();
                fStream.getFD().sync();
                oStream.close();

                Log.v(Client.TAG,"Serialization success");
            } catch (Exception e) {
                Log.e(TAG, "Error on persisting credential store", e);
            }
            return null;
        }
    }
}

File: android-lib/src/main/java/com/kinvey/android/AndroidCredentialStore.kt
/*
 *  Copyright (c) 2016, Kinvey, Inc. All rights reserved.
 *
 * This software is licensed to you under the Kinvey terms of service located at
 * http://www.kinvey.com/terms-of-use. By downloading, accessing and/or using this
 * software, you hereby accept such terms of service  (and any agreement referenced
 * therein) and agree that you have read, understand and agree to be bound by such
 * terms of service and are of legal age to agree to such terms with Kinvey.
 *
 * This software contains valuable confidential and proprietary information of
 * KINVEY, INC and is subject to applicable licensing agreements.
 * Unauthorized reproduction, transmission or distribution of this file and its
 * contents is a violation of applicable laws.
 *
 */
package com.kinvey.android

import android.content.Context
import android.os.AsyncTask
import android.os.Build.VERSION
import android.os.Build.VERSION_CODES
import com.google.common.base.Preconditions
import com.kinvey.java.Logger
import com.kinvey.java.auth.Credential
import com.kinvey.java.auth.CredentialStore
import java.io.*
import java.util.*

/**
 * @author mjsalinger
 * @since 2.0
 */
class AndroidCredentialStore(context: Context) : CredentialStore {
    private var credentials: HashMap<String, Credential>
    private val appContext: Context
    /** {@inheritDoc}  */
    @Throws(IOException::class)
    override fun load(userId: String): Credential {
        return credentials[userId]!!
    }

    /** {@inheritDoc}  */
    @Throws(IOException::class)
    override fun store(userId: String, credential: Credential) {
        Preconditions.checkNotNull(credential, "credential must not be null")
        Preconditions.checkNotNull(userId, "userId must not be null")
        credentials[userId] = credential
        persistCredentialStore()
    }

    /** {@inheritDoc}  */
    override fun delete(userId: String) {
        credentials.remove(userId)
        persistCredentialStore()
    }

    private fun persistCredentialStore() {
        if (VERSION.SDK_INT >= VERSION_CODES.HONEYCOMB) {
            PersistCredentialStore(appContext, credentials).executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR)
        } else {
            PersistCredentialStore(appContext, credentials).execute()
        }
    }

    @Throws(ClassNotFoundException::class)
    private fun retrieveCredentialStore() {
        var fIn: FileInputStream? = null
        var objIn: ObjectInputStream? = null
        if (appContext.getFileStreamPath("kinveyCredentials.bin").exists()) {
            try {
                fIn = appContext.openFileInput("kinveyCredentials.bin")
                objIn = ObjectInputStream(fIn)
                credentials = objIn.readObject() as HashMap<String, Credential>
            } catch (ex: IOException) {
                Logger.WARNING("Corrupt credential store detected")
            } finally {
                try {
                    fIn?.close()
                    objIn?.close()
                } catch (ioe: IOException) {
                    Logger.WARNING("Could not clean up resources")
                }
            }
        } else {
            persistCredentialStore()
        }
    }

    private class PersistCredentialStore(val context: Context, val credentials: HashMap<String, Credential>)
        : AsyncTask<Void?, Void?, Void?>() {
        override fun doInBackground(vararg args: Void?): Void? {
            try {
                val fStream: FileOutputStream = context.openFileOutput("kinveyCredentials.bin", Context.MODE_PRIVATE)
                val oStream = ObjectOutputStream(fStream)
                oStream.writeObject(credentials)
                oStream.flush()
                fStream.fd.sync()
                oStream.close()
                Logger.INFO("Serialization success")
            } catch (e: Exception) {
                Logger.ERROR("Error on persisting credential store")
            }
            return null
        }
    }

    companion object {
        private val TAG = AndroidCredentialStore::class.java.simpleName
    }

    init {
        appContext = context.applicationContext
        credentials = HashMap()
        try {
            retrieveCredentialStore()
        } catch (ex: ClassNotFoundException) {
            credentials = HashMap()
            persistCredentialStore()
            throw AndroidCredentialStoreException("Credential store corrupted and was rebuilt")
        }
    }
}
